theano/gof/tests/test_vm.py

```diff
@@ -371,3 +374,48 @@ def check_storage(storage_map):
         assert check_storage(storage_map)[0]
         assert len(set([id(v) for v in
                         storage_map.values()])) < len(storage_map)
+
+
+def test_shape_reallocation():
+        a, b, c, d = [tensor.dvector(n) for n in ['a', 'b', 'c', 'd']]
+        z = 3*a + b
+        x = c + 4*d
+        y = z + x
+
+        m = theano.compile.get_mode(theano.Mode(linker='vm_nogc'))
+        m_1 = m.excluding('fusion', 'inplace')
+        m_2 = m_1.excluding('ShapeOpt')
+
+        m_dict = {m_1: True, m_2: False}
```

Comment:
That is good. I think we should also test with linker "vm". Check the current implementation of test_reallocation. So we will end up with 4 element in the dict, the 2 actual + the same allow_gc=True. Don't use directly linker=vm, we need to explicitly disable the lazy with lazy=False. Otherwise, if people run the tests with the Theano flags vm.lazy=True, the test will fail.